fossas/spectrometer
# Add support for Yarn v2 projects

# Overview

This PR adds support for the v2 yarn lockfile format. When analyzing yarn projects, we first attempt to analyze `yarn.lock` as a v1 lockfile, then fall back to analyzing as a v2 lockfile.

Developer documentation for the yarn v2 lockfile is [rendered here](https://github.com/fossas/spectrometer/blob/yarn-v2/devdocs/buildtools/yarnv2.org)

## Acceptance criteria

- `fossa analyze` supports projects containing a yarn v2 lockfile, including all standard dependency types contained therein (see associated devdocs and `Strategy.Yarn.V2.Resolvers` module for more info)
- `fossa analyze` continues to support projects containing a yarn v1 lockfile

## Testing plan

- Automated unit tests, including an end-to-end test on the lockfile in the devdocs
- Manual sanity check tests on various yarn projects:
  1. Analyze a yarn v1 project in `--output` mode: `fossa analyze --output`
  2. Set the yarn version to v2: `yarn set version berry`
  3. Update the lockfile: `yarn install`
  4. Analyze the same project in `--output` mode: `fossa analyze --output`
  5. The dependency output from steps 1 and 4 should be identical (but with the addition of dependencies marked as "direct". Thanks yarn v2!)

## Risks

None

## References

Closes https://github.com/fossas/team-analysis/issues/526

## Checklist

Remaining:
- Convert devdocs to markdown
